source/common/http/filter_manager.cc

```diff
@@ -1264,6 +1266,15 @@ bool ActiveStreamDecoderFilter::recreateStream() {
   parent_.stream_info_.setResponseCodeDetails(
       StreamInfo::ResponseCodeDetails::get().InternalRedirect);
 
+  if (headers != nullptr) {
+    // Copy headers to ensure original headers are not modified during
+    // setResponseHeaders. This call is needed to ensure that the headers are
```

Comment:
Looking at setResponseHeaders it doesn't look like the headers can be modified during that call - can you update this comment to reflect what would happen without the copy?

Reply:
Ah, my comment is misleading actually. I'll fix it.

The copy isn't for modification, but for ownership. setResponseHeaders expects to take ownership of the headers (std::move), but these same underlying headers could potentially be std::move'd later in the flow in encodeHeaders, so we can't pass ownership as part of setupRedirect. That means we have to pass just a pointer until this point, and then convert that to a ResponseHeaderMapPtr&& by making an intermediate copy of the headers.

Is there a better way to do this?

Reply:
hm, @snowp has looked a lot into header ownership lately, so may have suggestions.  I think barring ideas from him I'm fine with the copy and a cleared up comment :-)

Reply:
Bump on this. @snowp do you have any suggestions here?

Reply:
I think the way I would avoid the copy would be to have onUpstreamHeaders to move the headers into the FM earlier in the function (like how the HCM does for request headers), as that allows us to reference the response headers from the FM throughout and not worry about when the ownership transfer happens. It also has the added benefit of being a bit more fool proof around handling cases where we end a stream before accepting a response (but have seen one), e.g. for hedged requests. 

That said the copy is probably fine if changing the ownership semantics ends up being too tricky

Reply:
Hey @snowp, I think there are two ways we can get the ownership straight, but I cannot decide which is better:

a) Make recreateStream return a ResponseHeaderMap&& to pass the ownership back when it fails to recreate the stream. We need to have a different way to return whether it successfully recreate the stream of course (a bool* in the parameter? return a std::pair<bool, ResponseHeaderMap>?)
b) Have StreamDecoderFilterCallbacks expose setResponseHeaders() and responseHeaders() so that we can set the headers earlier and refer to them after the ownership is passed

I am slightly leaning towards a), but would like to hear your thoughts.